nav2_util/include/nav2_util/service_client.hpp

```diff
@@ -78,9 +90,23 @@ class ServiceClient
       service_name_.c_str());
     auto future_result = client_->async_send_request(request);
 
-    if (callback_group_executor_.spin_until_future_complete(future_result, timeout) !=
-      rclcpp::FutureReturnCode::SUCCESS)
-    {
+    bool success = true;
+    if (use_default_callback_group_) {
+      std::future_status status;
```

Comment:
Why doesn't the `if` statement just mirror the `else` statement's logic, just using the `node_` to spin rather than the `callback_group_executor_`? You should also still use the timeout.

Reply:
we can't use `rclcpp::spin_until_future_complete(node_,future_result,timeout)` , look at https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/executors.hpp#L103-L112, it will create SingleThreadedExecutor . 
when `use_default_callback_group` is true, the default callback group will spin in thread, and we only need wait to use invoke.

Reply:
we can't use rclcpp::spin_until_future_complete(node_,future_result,timeout) , look at https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/executors.hpp#L103-L112, it will create SingleThreadedExecutor .
when use_default_callback_group is true, the default callback group will spin in thread, and we only need wait to use invoke.

Reply:
This is indeed something I had not thought about before. I did a brief survey of Nav2's use of spin until future is complete and was shocked at the results https://github.com/ros-planning/navigation2/search?p=1&q=spin_until_future_complete. Looking at this list **its not used in Nav2 at runtime**, all places are tests, python utilities, or the rviz panel. The rest have specified executors. That's a good sign, but also really unexpected. 

---

Thinking back to how we got here, this was under the impression that we could spin our input node rather than needing to create a new callback group to `spin_until_future_complete`. If we don't actually need to spin anything (e.g. we can just do some logic based on the future itself) then the entire idea of having a separate executor / callback group is largely moot, isn't it? I'm not 100% sure under the hood how it is that the future would actually get the updates in the service client with the above code if you never spin though and the node isn't spinning in the background, which was the core assumption of this method. If the services are in different processes, there's no possible way to have the client receive its result without spinning somewhere.

--- 

Also, I think you're looking at a different implementation of `spin_until_future_complete`. Here's the version I think you should be using in this case https://github.com/ros2/rclcpp/blob/6806cdf825d9fcdbb4bfe2c6e2a74c4cb8bc5e11/rclcpp/include/rclcpp/executor.hpp#L332. This input is the **future** not the node or executor. That future happens to be **based** on that node (and within it it calls `spin_once_impl(timeout_left);` which will spin on the main node)

Reply:
Good question though - that really did require me to dig in pretty deep. I think the right answer is to use the `spin_until_future_complete` API I linked to, I think that solves all of the issues and you can use my initial suggestion of making them more similar. That implementation largely does what you suggested in this PR but more robustly and includes the spin inside of the function.

Reply:
we must use a executor to spin, and the `spin_until_future_complete` API you [linked](https://github.com/ros2/rclcpp/blob/6806cdf825d9fcdbb4bfe2c6e2a74c4cb8bc5e11/rclcpp/include/rclcpp/executor.hpp#L332. ) is a function of `Class Executor`. so we can't avoid creating new  `SingleThreadedExecutor` to spin (`rclcpp::spin_until_future_complete()` will create a new `SingleThreadedExecutor` for each invoke).  

if we want to avoid creating new  `SingleThreadedExecutor`, only when the node has been spinning somewhere. but this will confuse developers (if we don't spin background, the `future` won't update and `invoke()` will block forever).

I can't find a better method than the previous `ServiceClient`, maybe this PR need be closed.

Reply:
Ah, looks like I made a goof, that's in the executor object not the global namespace. 

I agree, maybe we close this for now and document that this is something we'd like to revisit down the line.